design/proposal-zeroization.md

```diff
@@ -0,0 +1,91 @@
+# Proposal: Zeroization of Secrets after usage
+
+The goal is to minimize the lifetime of a Secret (and it's footprint) in RAM. 
+In order to achieve this goal, Secretless must:
+
+1. Be able to zeroize the original Secret and any deratives thereof. This is from the standpoint of technical feasibility of zeroing data types.
+2. Once a Secret has been used for it's intended purpose, carry out the zeroization.
+
+Ideally, 2 would happen in a centralized fashion to avoid the need for repeated boilerplate at the site of each derivative.
+
+
+## Journey of Secrets
+It is worth noting the journey of Secrets in Secretless.
+
+```
+Resolver.#Resolve -> Provider.#GetValue |-> Handler -> Listener -> Target Backend
+                                        |
+					|-> EventNotifier.#ResolveSecret
+```
+## Considerations
+
+Below are some of the considerations made in coming together with this proposal. Some of these need to be validated.
+
+1. Secrets arrive from the wire as byte slices. This is the only copy of the Secret in memory. If this is not the case then we'll need to think of ways to address this separately.
+2. Byte slices are passed by reference from place to place, unless resized or concatenated. Avoid uncontrolled resizes at all cost - we don't know what happens to the memory of the byte slice on resize. If we need to concatenate byte slices we can use the following utility function `ConcatBytes` and recognize the result as a derivative:
+```
+func ConcatBytes(bss... []byte) []byte {
+	size := 0
+	for _, bs := range bss  {
+		size += len(bs)
+	}
+
+	newByteSlice := make([]byte, size)
+
+	// avoided using append because every time the byte slice is resized
+	// memory will be copied over to the new byte slice
+	// and i guess the old byte slice will be GC'd - which isn't good enough
+	i := 0
+	for _, bs := range bss  {
+		for _, byte := range bs {
+			newByteSlice[i] = byte
+			i += 1
+		}
+	}
+	return newByteSlice
+}
+```
+3. We need to mindful of the passage of the Secret deravitives through the libraries responsible for using the Secret derivatives to establish a connection. e.g. Does the http listener create additional Secret deravitives, or is it enough to clean up the ones we pass to it?
```

Comment:
This problem is very, _very_ difficult.

The problem is that _even if_ we do a thorough code review of any library we use to which we will pass a secret, and _even if_  we do that so carefully that we don't miss anything, we are still vulnerable any time the library updates.  So the code reviews need to be baked into our package update process, and we have to lock packages to the reviewed version and never upgrade, or only upgrade after we have _re-reviewed_ the code.

Reply:
Sounds good. Of note, I don't think we tend to have external dependencies touch the secrets. The main concern will be code in the Go standard library, which gives me some hope :D